src/zenml/utils/template_utils.py

```diff
@@ -0,0 +1,181 @@
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Utilities for run templates."""
+
+from enum import Enum
+from typing import Any, Dict
+
+from pydantic import create_model
+
+from zenml.client import Client
+from zenml.config import ResourceSettings
+from zenml.config.base_settings import BaseSettings
+from zenml.config.pipeline_run_configuration import PipelineRunConfiguration
+from zenml.config.step_configurations import StepConfigurationUpdate
+from zenml.enums import StackComponentType
+from zenml.logger import get_logger
+from zenml.models import (
+    FlavorFilter,
+    PipelineDeploymentResponse,
+    StackResponse,
+)
+from zenml.stack import Flavor
+
+logger = get_logger(__name__)
+
+
+def generate_config_template(
+    deployment: PipelineDeploymentResponse,
+) -> Dict[str, Any]:
+    """Generate a run configuration template for a deployment.
+
+    Args:
+        deployment: The deployment.
+
+    Returns:
+        The run configuration template.
+    """
+    steps_configs = {
+        name: step.config.model_dump(
+            include=set(StepConfigurationUpdate.model_fields),
+            exclude={"name", "outputs"},
+        )
+        for name, step in deployment.step_configurations.items()
+    }
+
+    for config in steps_configs.values():
+        config["settings"].pop("docker", None)
+
+    pipeline_config = deployment.pipeline_configuration.model_dump(
+        include=set(PipelineRunConfiguration.model_fields),
+        exclude={"schedule", "build", "parameters"},
+    )
+
+    pipeline_config["settings"].pop("docker", None)
+
+    config_template = {
+        "run_name": deployment.run_name_template,
+        "steps": steps_configs,
+        **pipeline_config,
+    }
+    return config_template
+
+
+def generate_config_schema(
+    deployment: PipelineDeploymentResponse, stack: StackResponse
```

Comment:
Doesn't the deployment response already include the stack response?

Reply:
Yes, but not hydrated. This is actually a bigger thing I noticed. I also thought about this, and I think the best way would actually be to update these methods to work with a `DeploymentSchema` instead to avoid multiple DB calls.

There are actually many improvements here:
- Even with the `StackResponse` being passed in here, I still need to call the ZenStore again for each flavor of the stack components, which is lots of unnecessary DB calls.
- All our models have `Client().zen_store.get_something` in their `get_hydrated_version()` implementation. This means when we (accidently) hydrate a model on the server, it uses the `Client` which is something we should avoid. Apparently this is not failing at the moment, but I think we should look at that and fix it before it causes any issues.